### PR TITLE
Move deferred taint cleanup call to ensure all are removed

### PR DESCRIPTION
**Pull request by damemi**
**What type of PR is this?**
/kind flake

**What this PR does / why we need it**:
In some CI clusters, we've found unrelated tests failing due to taints found only in this test being present on nodes.

One example is `[sig-cluster-lifecycle][Feature:Machines][Serial] Managed cluster should grow and decrease when scaling different machineSets simultaneously [Suite:openshift/conformance/serial]`, which we have seen fail with `node/ci-op-4it1h40d-22d24-w4mls-worker-c-tc9qh had unexpected taints: []v1.Taint{v1.Taint{Key:"kubernetes.io/e2e-taint-key-90ae0008-a503-4dd3-baf9-b8b02123f7f7", Value:"testing-taint-value-1fe3e0bf-1727-44ba-97cb-c6900294e005", Effect:"PreferNoSchedule", TimeAdded:(*v1.Time)(nil)}}`

It seems that `addRandomTaintToNode` makes a ginkgo expect call, which I think if it fails causes a panic which prevents the following `defer e2enode.RemoveTaintOffNode` call from being pushed on the defer stack. This leaves just 1 taint, which is obviously enough to cause other tests to fail.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```
/sig scheduling

### CODE REVIEW

File: test/e2e/scheduling/priorities.go [hunk 1]
```diff
@@ -304,18 +306,18 @@ var _ = SIGDescribe("SchedulerPriorities [Serial]", func() {
 		nodeName := nodeList.Items[0].Name
 
 		ginkgo.By("Trying to apply 10 (tolerable) taints on the first node.")
+		// Defer cleaning up the taints to guarantee that all matching taints get removed after the test
+		defer cleanupTestTaints(cs, nodeName)
```
Comment (ahg-g) [inline test/e2e/scheduling/priorities.go:310]: to avoid replicating the remove taint code, we could create the taints here in a list

```
tolerableTaints := getRandomTaints(len(nodelList.Items))
intolerableTaints := getRandomTaints(len(nodelList.Items))
defer e2enode.RemoveTaintOffNode(cs, node.Name, tolerableTaints...)
defer e2enode.RemoveTaintOffNode(cs, node.Name, intolerableTaints...)
```

and then below just use those taints to apply to the nodes.

Comment (damemi) [reply test/e2e/scheduling/priorities.go:310]: I'm not sure what you mean by your code example, but we are only trying to add taints to 1 node (so `len(nodeList.Items)` shouldn't be necessary)

Do you want to refactor `addRandomTaintToNode` to just return the taints, instead of actually adding them? Or introduce a new function `getRandomTaints` that does that?

Also, `e2enode.RemoveTaintOffNode` only accepts 1 taint, so we could write a wrapper for it that takes a list.

Comment (ahg-g) [reply test/e2e/scheduling/priorities.go:310]: > I'm not sure what you mean by your code example, but we are only trying to add taints to 1 node (so len(nodeList.Items) shouldn't be necessary)

I see line 320 adds taints to the rest of the of the nodes.

> Do you want to refactor addRandomTaintToNode to just return the taints, instead of actually adding them? Or introduce a new function getRandomTaints that does that?

`getRandomTaint` would basically be:
```
func getRandomTaint() {
 return v1.Taint{
		Key:    fmt.Sprintf("kubernetes.io/scheduling-priorities-e2e-taint-key-%s", string(uuid.NewUUID())),
		Value:  fmt.Sprintf("testing-taint-value-%s", string(uuid.NewUUID())),
		Effect: v1.TaintEffectPreferNoSchedule,
	}
}
```

addRandomTaintToNode becomes addTaintToNode:
```
func addRandomTaintToNode(cs clientset.Interface, nodeName string, taint* v1.Taint) {
	e2enode.AddOrUpdateTaintOnNode(cs, nodeName, *taint)
	framework.ExpectNodeHasTaint(cs, nodeName, taint)
}
```

> Also, e2enode.RemoveTaintOffNode only accepts 1 taint, so we could write a wrapper for it that takes a list.

ah, I was looking at the function under [staging/src/k8s.io/cloud-provider/node/helpers/taints.go](https://github.com/kubernetes/kubernetes/blob/cea1d4e20b4a7886d8ff65f34c6d4f95efcb4742/staging/src/k8s.io/cloud-provider/node/helpers/taints.go#L147), yeah we could wrap it.

Comment (damemi) [reply test/e2e/scheduling/priorities.go]: Added the wrapper to pass a list of taints, I agree that it's best to avoid duplicating the removal code.

For generating the random taints, I went ahead and made a list for the tolerable taints on the first node + a map for each remaining node's intolerable taints. Let me know what you think

File: test/e2e/scheduling/priorities.go [hunk 2]
```diff
@@ -303,19 +303,40 @@ var _ = SIGDescribe("SchedulerPriorities [Serial]", func() {
 		// Apply 10 taints to first node
 		nodeName := nodeList.Items[0].Name
 
-		ginkgo.By("Trying to apply 10 (tolerable) taints on the first node.")
```
Comment (ahg-g) [inline test/e2e/scheduling/priorities.go:306]: why was this removed?

File: test/e2e/scheduling/priorities.go [hunk 3]
```diff
@@ -582,13 +603,15 @@ func createRC(ns, rsName string, replicas int32, rcPodLabels map[string]string,
...
-		Key:    fmt.Sprintf("kubernetes.io/e2e-taint-key-%s", string(uuid.NewUUID())),
+func getRandomTaint() v1.Taint {
+	return v1.Taint{
+		Key:    fmt.Sprintf("kubernetes.io/scheduling-priorities-e2e-taint-key-%s", string(uuid.NewUUID())),
 		Value:  fmt.Sprintf("testing-taint-value-%s", string(uuid.NewUUID())),
 		Effect: v1.TaintEffectPreferNoSchedule,
 	}
+}
+
+func addRandomTaintToNode(cs clientset.Interface, nodeName string, testTaint v1.Taint) {
```
Comment (ahg-g) [inline test/e2e/scheduling/priorities.go:614]:
```suggestion
func addTaintToNode(cs clientset.Interface, nodeName string, testTaint v1.Taint) {
```

**Review by ahg-g — COMMENTED** [review]
looks good

File: test/e2e/scheduling/priorities.go (hunk 2 above)
Comment (damemi) [reply test/e2e/scheduling/priorities.go:306]: It was moved down to where those taints are actually applied (L330)

**Review by damemi — COMMENTED** [review]
.

File: test/e2e/scheduling/priorities.go (hunk 2 above)
Comment (ahg-g) [reply test/e2e/scheduling/priorities.go:306]: sg

**Comment by damemi** [comment]
/cc @spiffxp 
ptal, for sig-testing approval

**Comment by ahg-g** [comment]
/lgtm 

for scheduling

File: test/e2e/scheduling/priorities.go [hunk 4]
```diff
@@ -303,19 +303,40 @@ var _ = SIGDescribe("SchedulerPriorities [Serial]", func() {
 		// Apply 10 taints to first node
 		nodeName := nodeList.Items[0].Name
 
-		ginkgo.By("Trying to apply 10 (tolerable) taints on the first node.")
+		// First, create a set of tolerable taints (+tolerations) for the first node.
+		// We immediately defer the removal of these taints because addTaintToNode can
+		// panic and RemoveTaintsOffNode does not return an error if the taint does not exist.
```
Comment (spiffxp) [inline test/e2e/scheduling/priorities.go:308]: I would move the defer calls and this comment explaining why to right before the calls to `addTaintToNode`, there's nothing about generating them that's going to require a defer

File: test/e2e/scheduling/priorities.go [hunk 5]
```diff
@@ -303,19 +303,40 @@ var _ = SIGDescribe("SchedulerPriorities [Serial]", func() {
...
+			for i := 0; i < 10; i++ {
+				nodeTaints = append(nodeTaints, getRandomTaint())
+			}
+			intolerableTaints[nodeList.Items[i].Name] = nodeTaints
+			defer e2enode.RemoveTaintsOffNode(cs, nodeList.Items[i].Name, nodeTaints)
+		}
+
+		// Apply the tolerable taints generated above to the first node
+		ginkgo.By("Trying to apply 10 (tolerable) taints on the first node.")
+		for _, taint := range tolerableTaints {
```
Comment (spiffxp) [inline test/e2e/scheduling/priorities.go:331]: e.g.
```suggestion
		ginkgo.By("Trying to apply 10 (tolerable) taints on the first node.")
		defer e2enode.RemoveTaintsOffNode(cs, nodeName, tolerableTaints)
		for _, taint := range tolerableTaints {
```

File: test/e2e/scheduling/priorities.go [hunk 6]
```diff
@@ -303,19 +303,40 @@ var _ = SIGDescribe("SchedulerPriorities [Serial]", func() {
...
+		}
+		// Apply the intolerable taints to each of the following nodes
 		ginkgo.By("Adding 10 intolerable taints to all other nodes")
 		for i := 1; i < len(nodeList.Items); i++ {
 			node := nodeList.Items[i]
-			for i := 0; i < 10; i++ {
-				testTaint := addRandomTaintToNode(cs, node.Name)
-				defer e2enode.RemoveTaintOffNode(cs, node.Name, *testTaint)
+			for _, taints := range intolerableTaints[node.Name] {
+				addTaintToNode(cs, node.Name, taints)
```
Comment (spiffxp) [inline test/e2e/scheduling/priorities.go:339]: e.g.
```suggestion
			taints := intolerableTaints[node.Name]
			defer e2enode.RemoveTaintsOffNode(cs, node.Name, taints)
			for _, taint := range taints {
				addTaintToNode(cs, node.Name, taint)
```

File: test/e2e/scheduling/priorities.go [hunk 7]
```diff
@@ -582,13 +603,15 @@ func createRC(ns, rsName string, replicas int32, rcPodLabels map[string]string,
 	return rc
 }
 
-func addRandomTaintToNode(cs clientset.Interface, nodeName string) *v1.Taint {
-	testTaint := v1.Taint{
-		Key:    fmt.Sprintf("kubernetes.io/e2e-taint-key-%s", string(uuid.NewUUID())),
+func getRandomTaint() v1.Taint {
+	return v1.Taint{
+		Key:    fmt.Sprintf("kubernetes.io/scheduling-priorities-e2e-taint-key-%s", string(uuid.NewUUID())),
```
Comment (spiffxp) [inline test/e2e/scheduling/priorities.go:608]: Is there a max taint key length we should be watching out for?  Labels are capped at 63; the hardcoded string part here is 52 chars and `string(uuid.NewUUID())` appears to be 37 chars

**Review by spiffxp — CHANGES_REQUESTED** [review]
Some suggestions and a question

File: test/e2e/scheduling/priorities.go (hunk 7 above)
Comment (spiffxp) [reply test/e2e/scheduling/priorities.go:608]: https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#taint

>The key must begin with a letter or number, and may contain letters, numbers, hyphens, dots, and underscores, up to 253 characters.

ok, this is well within the limit

**Comment by damemi** [comment]
@spiffxp sorry for the delay, I just updated with your feedback. ptal, thanks!

File: test/e2e/scheduling/priorities.go [hunk 8]
```diff
@@ -303,19 +303,41 @@ var _ = SIGDescribe("SchedulerPriorities [Serial]", func() {
...
 		}
+		// Generate 10 intolerable taints for each of the remaining nodes
+		intolerableTaints := make(map[string][]v1.Taint)
+		for i := 1; i < len(nodeList.Items); i++ {
+			nodeTaints := make([]v1.Taint, 0)
+			for i := 0; i < 10; i++ {
+				nodeTaints = append(nodeTaints, getRandomTaint())
+			}
+			intolerableTaints[nodeList.Items[i].Name] = nodeTaints
+			defer e2enode.RemoveTaintsOffNode(cs, nodeList.Items[i].Name, nodeTaints)
```
Comment (spiffxp) [inline test/e2e/scheduling/priorities.go:323]: I think this was missed in the edit; you're doing this below (as I suggested)
```suggestion
```

File: test/e2e/scheduling/priorities.go (hunk 6 above)
Comment (spiffxp) [reply test/e2e/scheduling/priorities.go:340]: if not the intermediate variable approach I suggested above, I would at least like you to drop the plural from the variable you're using in the for loop; `taints` keeps making me think you're adding a list, but `addTaintToNode` only takes a singular taint, so it looks like there's a bug, but the compiler's happy, and.... anyway please save future maintainers the confusion
```suggestion
			for _, taint := range intolerableTaints[node.Name] {
				addTaintToNode(cs, node.Name, taint)
```

**Review by spiffxp — CHANGES_REQUESTED** [review]
Sorry for _my_ delay in reviewing, you're almost there

File: test/e2e/scheduling/priorities.go (hunk 8 above)
Comment (damemi) [reply test/e2e/scheduling/priorities.go]: Thanks, missed that

File: test/e2e/scheduling/priorities.go (hunk 6 above)
Comment (damemi) [reply test/e2e/scheduling/priorities.go]: I see the confusion, updated to just remove the pluralization

**Comment by damemi** [comment]
Bump, is this all set now? :)

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm

File: test/e2e/scheduling/priorities.go (hunk 7 above)
Comment (damemi) [reply test/e2e/scheduling/priorities.go:608]: @spiffxp did that limit recently increase? Trying to port this to openshift I am seeing an error that this key is too long: `name part must be no more than 63 characters`

Comment (aojea) [reply test/e2e/scheduling/priorities.go:608]: @damemi this is failing in the upstream CI too, due to this key limitation
https://testgrid.k8s.io/sig-network-gce#gci-gce-serial&width=5
```
                       Type: "FieldValueInvalid",
                        Message: "Invalid value: \"kubernetes.io/scheduling-priorities-e2e-taint-key-c19ab4b9-07b1-4a86-9a0f-52c1a6028392\": name part must be no more than 63 characters",
                        Field: "metadata.taints[0].key",
                    },
                ],
```

Comment (aojea) [reply test/e2e/scheduling/priorities.go:608]: ping @spiffxp ^^^

Comment (BenTheElder) [reply test/e2e/scheduling/priorities.go:608]: filed https://github.com/kubernetes/kubernetes/pull/99020

we may also need to update the docs.
